agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -141,108 +184,129 @@ public Function thenSort(@NonNull final Comparator comparator) {
     return createFunction();
   }
 
-  private static final class LimitFunction<T> implements Function<List<T>, List<T>> {
-    private final int limit;
-
-    LimitFunction(final int limit) {
-      this.limit = limit;
-    }
-
-    @NonNull
-    @Override
-    public List<T> apply(@NonNull final List<T> input) {
-      if (input.size() < limit) {
-        return input;
-      }
-      if (limit <= 0) {
-        return emptyList();
-      }
-      return new ArrayList<>(input.subList(0, limit));
-    }
-  }
-
-  private static final class MapFunction<F, T> implements Function<List<F>, List<T>> {
-    @NonNull
-    private final Function<F, T> function;
-
-    MapFunction(@NonNull final Function<F, T> function) {
-      this.function = checkNotNull(function);
-    }
-
-    @NonNull
-    @Override
-    @SuppressWarnings("unchecked")
-    public List<T> apply(@NonNull final List<F> input) {
-      if (input.isEmpty()) {
-        return emptyList();
-      }
-      final List<T> result = new ArrayList(input.size());
-      for (final F item : input) {
-        result.add(function.apply(item));
-      }
-      return result;
-    }
+  @NonNull
+  @Override
+  public Function thenSort() {
+    sort();
+    return createFunction();
   }
 
-  private static final class ChainFunction implements Function {
+  private static final class CompiledFunction implements Function {
     @NonNull
-    private final Function[] functions;
+    private final Object[] directives;
 
-    ChainFunction(@NonNull final Function[] functions) {
-      this.functions = checkNotNull(functions);
+    CompiledFunction(@NonNull final Object[] directives) {
+      this.directives = checkNotNull(directives);
     }
 
     @NonNull
     @Override
     @SuppressWarnings("unchecked")
     public Object apply(@NonNull final Object input) {
       Object item = input;
-      for (final Function function : functions) {
-        item = function.apply(item);
+      int i = 0;
+      final int length = directives.length;
+      while (i < length) {
+        final Object type = directives[i++];
+        final Object command = directives[i++];
+        if (type == APPLY) {
+          item = ((Function) command).apply(item);
+        } else if (type == MAP) {
+          final List list = (List) item;
+          final int size = list.size();
+          if (size <= 0) {
+            item = emptyList();
+          } else if (list instanceof FunctionCompilerList) {
```

Comment:
unfortunately you cannot do it, it's possible that result of one compiled function will be saved and passed to some other compiled functions (moe then one/once). You shouldn't change the original input list.